packages/suite-desktop/src-electron/modules/coinjoin.ts

```diff
@@ -0,0 +1,102 @@
+
+const SERVICE_NAME = '@trezor/coinjoin';
+
+const init: Module = ({ mainWindow }) => {
+    const { logger } = global;
+
+    const backends: CoinjoinBackend[] = [];
+    const clients: CoinjoinClient[] = [];
+
+    logger.debug(SERVICE_NAME, `Starting service`);
+
+    const backendProxyOptions: IpcProxyHandlerOptions<CoinjoinBackend> = {
+        onCreateInstance: (settings: ConstructorParameters<typeof CoinjoinBackend>[0]) => {
+            const backend = new CoinjoinBackend(settings);
+            backends.push(backend);
+            return {
+                onRequest: (method, _params) => {
+                    logger.debug(SERVICE_NAME, `CoinjoinBackend call ${method}`);
+                    return backend[method].bind(backend); // bind method to instance context
+                },
+                onAddListener: (eventName, listener) => {
+                    logger.debug(SERVICE_NAME, `CoinjoinBackend add listener ${eventName}`);
+                    backend.on(eventName, listener);
+                },
+                onRemoveListener: eventName => {
+                    logger.debug(SERVICE_NAME, `CoinjoinBackend remove listener ${eventName}`);
+                    backend.removeAllListeners(eventName);
+                    return backend;
+                },
+            };
+        },
+        debug: console,
+    };
+
+    const clientProxyOptions: IpcProxyHandlerOptions<CoinjoinClient> = {
+        onCreateInstance: (settings: ConstructorParameters<typeof CoinjoinClient>[0]) => {
+            const client = new CoinjoinClient(settings);
+            clients.push(client);
+            return {
+                onRequest: (method, params) => {
+                    logger.debug(SERVICE_NAME, `CoinjoinClient call ${method}`);
+                    if (method === 'enable') {
+                        return async () => {
+                            logger.debug(SERVICE_NAME, `CoinjoinClient enable ${params}`);
+                            const response = await client[method](...params);
+                            // TODO: start coinjoin middleware binary
+                            return response;
+                        };
+                    }
+                    return client[method].bind(client); // bind method to instance context
+                },
+                onAddListener: (eventName, listener) => {
+                    logger.debug(SERVICE_NAME, `CoinjoinClient add listener ${eventName}`);
+                    client.on(eventName, listener);
+                },
+                onRemoveListener: eventName => {
+                    logger.debug(SERVICE_NAME, `CoinjoinClient remove listener ${eventName}`);
+                    client.removeAllListeners(eventName);
+                    return client;
+                },
+            };
+        },
+        debug: console,
+    };
+
+    return () => {
+        const unregisterBackendProxy = createIpcProxyHandler<CoinjoinBackend>(
```

Comment:
The generic parameter here is probably unnecessary as it could be inferred.

Reply:
right, removed